macaw-aarch32/src/Data/Macaw/ARM/Arch.hs

```diff
@@ -137,25 +133,60 @@ instance TF.FunctorF ARMTermStmt where
 instance TF.TraversableF ARMTermStmt where
   traverseF go tstmt =
     case tstmt of
-      ARMSyscall imm -> pure (ARMSyscall imm)
-      ThumbSyscall imm -> pure (ThumbSyscall imm)
       ReturnIf cond -> ReturnIf <$> go cond
       ReturnIfNot cond -> ReturnIfNot <$> go cond
 
 rewriteTermStmt :: ARMTermStmt (MC.Value ARM.AArch32 src) -> Rewriter ARM.AArch32 s src tgt (ARMTermStmt (MC.Value ARM.AArch32 tgt))
 rewriteTermStmt s =
     case s of
-      ARMSyscall imm -> pure $ ARMSyscall imm
-      ThumbSyscall imm -> pure (ThumbSyscall imm)
       ReturnIf cond -> ReturnIf <$> rewriteValue cond
       ReturnIfNot cond -> ReturnIfNot <$> rewriteValue cond
 
 -- ----------------------------------------------------------------------
 -- ARM functions.  These may return a value, and may depend on the
--- current state of the heap and the set of registeres defined so far
+-- current state of the heap and the set of registers defined so far
 -- and the result type, but should not affect the processor state.
 
 data ARMPrimFn (f :: MT.Type -> Type) tp where
+  -- | Issue a system call
+  --
+  -- The intent is that the user provides a mapping from system call numbers to
+  -- handlers in macaw-aarch32-symbolic, enabling the translation to crucible to
+  -- replace this operation with a lookup + call to a function handle.  By
+  -- capturing all of the necessary registers as inputs and outputs here,
+  -- uniform treatment is possible.  See the x86 version for a more detailed
+  -- account of the translation strategy.
+  --
+  -- The 'WI.W' is the immediate operand embedded in the opcode
+  --
+  -- The rest of the arguments are all of the registers that participate in the
+  -- syscall protocol (at least on Linux). Arguments are passed in r0-r6, while
+  -- the syscall number is in r7.
+  --
+  -- The system call can return up to two values (in r0 and r1)
+  ARMSyscall :: WI.W 24
```

Comment:
Aside from the immediate operand, the types of `ARMSyscall` and `ThumbSyscall` are identical. Would it be cleaner to combine the two data constructors into one and have a separate data type for the immediate operand? Perhaps something like:

```hs
data ARMPrimFn (f :: MT.Type -> Type) tp where
  Syscall :: SyscallImm
          -> f (MT.BVType 32)
          -> f (MT.BVType 32)
          -> f (MT.BVType 32)
          -> f (MT.BVType 32)
          -> f (MT.BVType 32)
          -> f (MT.BVType 32)
          -> f (MT.BVType 32)
          -> f (MT.BVType 32)
          -> ARMPrimFn f (MT.TupleType [MT.BVType 32, MT.BVType 32])
  ...

data SyscallImm
  = ARMImm (WI.W 24)
  | ThumbImm (WI.W 8)
```

This would help avoid some noticeable code duplication involving the `*Syscall` data constructors (e.g., in `aarch32GenFn`). I'm not sure if this would work out to be less code overall, since there are other parts of the code that do different things for ARM and Thumb, but the idea seems tantalizing.

Reply:
Yeah this was bugging me too. I think I might just zero extend the immediate and use the same constructor for both. The kernel can't actually tell which instruction got you there.